### DISCUSSION TITLE
Unexpected Tensor shapes within Keras Model while using Sionna's RootRaisedCosineFilters

### DISCUSSION THREAD

**Discussion by kapis20** [NVlabs/sionna#671] [Q&A] [answered]
Hi while building the end-2-end communication layer, I encountered an unexpected shape change of Tensors's shape. Below is an instance of my two filters:

![image](https://github.com/user-attachments/assets/5bae9b6d-c42a-4ecd-a1f7-c292d3f8f38b)

The problem occurs when the offset is introduced in the downsampling layer to compensate for the delay. 
Here is a detailed explanation of the problem: 

1. **After upsampling**: The Tensor shape is as expected, [BatchSize, M*n], where 

- M = samples per symbol (sampling factor) 

- n = number of symbols 

2.  **After filtering**: The shape changes and some unexpected shapes appear (differs from a benchmark/test script). I have tried fixing this by applying padding to ensure the shape remains consistent as [BatchSize, M*n].

3. **After Downsampling**: The Tensor shape changes in an unexpected way (again differs from the benchmark). The new shape of Tensor is: [BatchSize, x], where x<n. 

4. **Tried potential solutions** : I have tried fixing it by adding manual padding, to match the shape of [BatchSize,n], but because it is executed in graph mode (see below), then I failed to do so successfully. 

![image](https://github.com/user-attachments/assets/95285148-e9c7-4149-a82e-2d6bd2b60da8)

Interestingly, when benchmarking using Sionna’s QAM source without Keras models, the issue does not occur. In those cases, regardless of the attributes used, the downsampling step always yields the expected shape of [BatchSize, n]
Benchmark script: 
![image](https://github.com/user-attachments/assets/08708998-703e-4f95-8e2a-45fe8fbfbc30)

5. **Resolution**: The issue was resolved when applying padding to the receiver signal only 
![image](https://github.com/user-attachments/assets/3eebd47e-1f54-4f56-adb1-6eb1b3101f1a)

6. **Observations** I believe the issue I’m encountering is related to rounding in Keras layers, which leads to signal truncation. However, I am struggling to fully understand why this behavior occurs. Here’s a breakdown of the problem:

- [ ] **Padding Behaviour**:

- When I apply padding to both the transmitter and receiver filters, the issue persists, and the tensor shape is unexpected.
- However, if I apply padding only to the receiver filter, everything works as expected, and the tensor shape remains consistent.

- [ ] Unexpected Behavior:
- This behaviour differs from my benchmarks, where I do not encounter the same issue. however I do not manipulate tensors in that script and it is not executed in the graph mode, so everything is calculated in Python. 

- [ ] Lack of Clarity in Keras Layer Behaviour:
- It seems that the way tensors are manipulated in Keras layers is different from the manual implementations used in my benchmarks. This makes it hard for me to pinpoint the exact reason why applying padding to both filters does not resolve the issue.
- Additionally, debugging the code in graph mode is challenging because tensors are not easily accessible within the model. This further complicates understanding the underlying behaviour.

7. **My Questions**:
-Why does padding both filters fail to resolve the issue, while padding only the receiver filter works fine?
-Could this be due to the way Keras layers handle rounding, or is there another factor in how tensors are manipulated within Keras layers?

I’d appreciate any insights or guidance on understanding this behaviour, especially in the context of Keras layer tensor manipulation and rounding. Please let me know if you require more input into the problem.

**Comment by SebastianCa** [comment]
Hi @kapis20, 
can you please provide a small code snippet that allows to reproduce the error?

**Reply by kapis20** [reply, marked as answer]
Hi @SebastianCa, thank you for your reply! It turns out I passed the wrong parameter to Sionna's Downsampling class, which caused the unexpected shape of the Tensor. Initially, I omitted it, but after diligent evaluation of the same conventional model outside graph mode, the issue became much clearer. Everything works as expected! Thanks once again.
